fern/core/addons.js

```diff
@@ -0,0 +1,102 @@
+const stream = require("stream");
+const { promisify } = require("util");
+const path = require("path");
+const fs = require("fs");
+
+const got = require("got");
+const execa = require("execa");
+const Listr = require("listr");
+const readdir = require("recursive-readdir");
+
+const { getCacheDir } = require("./caching.js");
+const { fileExists, folderExists, ensureFolderExists } = require("./utils.js");
+
+const btoa = b => Buffer.from(b).toString('base64')
+
+function getAddonUrlHash(url) {
+  return btoa(url).slice(0, -1); // Remove trailing `=`
+}
+
+async function getPathToCachedAddon(addonName, url) {
+  const addonUrlHash = getAddonUrlHash(url);
+  const cache = await getCacheDir(addonName);
+  return path.join(cache, addonUrlHash);
+}
+
+async function getAddonId(addonName, url) {
+  const folder = await getPathToCachedAddon(addonName, url);
+  const manifestPath = path.join(folder, 'manifest.json');
+  const manifest = require(manifestPath);
+  return manifest['browser_specific_settings']['gecko']['id'];
+}
+
+async function use(addons) {
+  const tasks = await Promise.all(Object.keys(addons).map(async (addonName) => {
+    const cache = await getCacheDir(addonName);
+    const url = addons[addonName];
+    const folder = await getPathToCachedAddon(addonName, url);
+    const addonUrlHash = getAddonUrlHash(url);
+    const archive = path.join(cache, `${addonUrlHash}.zip`);
+
+    return [
+      {
+        title: `Download ${addonName}`,
+        skip: () => fileExists(archive),
+        task: () =>
+          promisify(stream.pipeline)(
```

Comment:
Did you find a way to catch errors here?

Reply:
Nope, but the error I was getting was caused by the ghostery-search-extension repository being private. Even if download step do not fail, then the unzip one does, so handing errors here is nice to have only.

Reply:
done